tile38-1.10.1 build broken under FreeBSD

I try to update tile38 from 1.10.0 to 1.10.1 and get the following error:

~~~
% ./build.sh 
# runtime
/usr/local/go/src/runtime/signal_unix.go:28:21: _SIG_DFL redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:15:21
/usr/local/go/src/runtime/signal_unix.go:29:21: _SIG_IGN redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:16:21
/usr/local/go/src/runtime/signal_unix.go:39:5: fwdSig redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:25:5
/usr/local/go/src/runtime/signal_unix.go:50:2: disableSigChan redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:35:2
/usr/local/go/src/runtime/signal_unix.go:51:2: enableSigChan redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:36:2
/usr/local/go/src/runtime/signal_unix.go:52:2: maskUpdatedChan redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:37:2
/usr/local/go/src/runtime/signal_unix.go:64:5: signalsOK redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:49:5
/usr/local/go/src/runtime/signal_unix.go:70:22: initsig redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:55:22
/usr/local/go/src/runtime/signal_unix.go:108:38: sigInstallGoHandler redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:90:37
/usr/local/go/src/runtime/signal_unix.go:136:20: sigenable redeclared in this block
        previous declaration at /usr/local/go/src/runtime/signal1_unix.go:115:20
/usr/local/go/src/runtime/signal_unix.go:136:20: too many errors
~~~


My env:
~~~
% uname -mrs
FreeBSD 12.0-CURRENT amd64
~~~

~~~
% go version
go version go1.9.2 freebsd/amd64
~~~

Previous releases (including 1.10.0)  did not have any problem.

--- a/internal/server/json.go
+++ b/internal/server/json.go
@@ -41,52 +41,80 @@ func jsonString(s string) string {
 	return string(b)
 }
 
-func isJsonNumber(s string) bool {
+func isJsonNumber(data string) bool {
 	// Returns true if the given string can be encoded as a JSON number value.
 	// See:
 	// https://json.org
 	// http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf
-	l := len(s)
-	if l == 0 {
+	if data == "" {
 		return false
 	}
 	i := 0
-	if s[i] == '-' {
+	// sign
+	if data[i] == '-' {
 		i++
 	}
-	switch {
-	case i == l:
+	if i == len(data) {
 		return false
-	case s[i] == '0': // 0 must not be followed by any digits.
+	}
+	// int
+	if data[i] == '0' {
 		i++
-		break
-	case s[i] >= '1' && s[0] <= '9':
-		for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ { // scan over digits
+	} else {
+		for ; i < len(data); i++ {
+			if data[i] >= '0' && data[i] <= '9' {
+				continue
+			}
+			break
 		}
-	default:
-		return false
 	}
-	if i == l {
+	// frac
+	if i == len(data) {
 		return true
 	}
-	if  l-i > 1 && s[i] == '.' {
-		for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ {
+	if data[i] == '.' {
+		i++
+		if i == len(data) {
+			return false
+		}
+		if data[i] < '0' || data[i] > '9' {
+			return false
+		}
+		i++
+		for ; i < len(data); i++ {
+			if data[i] >= '0' && data[i] <= '9' {
+				continue
+			}
+			break
 		}
 	}
-	if l-i > 1 && (s[i] == 'e' || s[i] == 'E') {
+	// exp
+	if i == len(data) {
+		return true
+	}
+	if data[i] == 'e' || data[i] == 'E' {
 		i++
-		if s[i] == '-' || s[i] == '+' {
+		if i == len(data) {
+			return false
+		}
+		if data[i] == '+' || data[i] == '-' {
 			i++
 		}
-		switch {
-		case i == l:
+		if i == len(data) {
 			return false
-		case s[i] >= '0' && s[0] <= '9':
-			for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ {
+		}
+		if data[i] < '0' || data[i] > '9' {
+			return false
+		}
+		i++
+		for ; i < len(data); i++ {
+			if data[i] >= '0' && data[i] <= '9' {
+				continue
 			}
+			break
 		}
 	}
-	return i == l
+	return i == len(data)
 }
 
 func appendJSONSimpleBounds(dst []byte, o geojson.Object) []byte {

--- a/internal/server/json.go
+++ b/internal/server/json.go
@@ -40,6 +40,55 @@ func jsonString(s string) string {
 	b[len(b)-1] = '"'
 	return string(b)
 }
+
+func isJsonNumber(s string) bool {
+	// Returns true if the given string can be encoded as a JSON number value.
+	// See:
+	// https://json.org
+	// http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf
+	l := len(s)
+	if l == 0 {
+		return false
+	}
+	i := 0
+	if s[i] == '-' {
+		i++
+	}
+	switch {
+	case i == l:
+		return false
+	case s[i] == '0': // 0 must not be followed by any digits.
+		i++
+		break
+	case s[i] >= '1' && s[0] <= '9':
+		for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ { // scan over digits
+		}
+	default:
+		return false
+	}
+	if i == l {
+		return true
+	}
+	if  l-i > 1 && s[i] == '.' {
+		for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ {
+		}
+	}
+	if l-i > 1 && (s[i] == 'e' || s[i] == 'E') {
+		i++
+		if s[i] == '-' || s[i] == '+' {
+			i++
+		}
+		switch {
+		case i == l:
+			return false
+		case s[i] >= '0' && s[0] <= '9':
+			for i++; i < l && s[i] >= '0' && s[i] <= '9'; i++ {
+			}
+		}
+	}
+	return i == l
+}
+
 func appendJSONSimpleBounds(dst []byte, o geojson.Object) []byte {
 	bbox := o.Rect()
 	dst = append(dst, `{"sw":{"lat":`...)
@@ -183,13 +232,7 @@ func (c *Server) cmdJset(msg *Message) (res resp.Value, d commandDetails, err er
 	if !str && !raw {
 		switch val {
 		default:
-			if len(val) > 0 {
-				if (val[0] >= '0' && val[0] <= '9') || val[0] == '-' {
-					if _, err := strconv.ParseFloat(val, 64); err == nil {
-						raw = true
-					}
-				}
-			}
+			raw = isJsonNumber(val)
 		case "true", "false", "null":
 			raw = true
 		}

--- a/internal/server/json_test.go
+++ b/internal/server/json_test.go
@@ -18,3 +18,29 @@ func BenchmarkJSONMarshal(t *testing.B) {
 		json.Marshal(s)
 	}
 }
+
+func TestIsJsonNumber(t *testing.T) {
+	test := func(expected bool, val string) {
+		actual := isJsonNumber(val)
+		if expected != actual {
+			t.Fatalf("Expected %t == isJsonNumber(\"%s\") but was %t", expected, val, actual)
+		}
+	}
+	test(false, "")
+	test(false, "-")
+	test(false, "foo")
+	test(false, "0123")
+	test(false, "1.")
+	test(false, "1.0e")
+	test(false, "1.0e-")
+	test(false, "1.0E10NaN")
+	test(false, "1.0ENaN")
+	test(true, "-1")
+	test(true, "0")
+	test(true, "0.0")
+	test(true, "42")
+	test(true, "1.0E10")
+	test(true, "1.0e10")
+	test(true, "1E+5")
+	test(true, "1E-10")
+}

--- a/tests/json_test.go
+++ b/tests/json_test.go
@@ -5,6 +5,8 @@ import "testing"
 func subTestJSON(t *testing.T, mc *mockServer) {
 	runStep(t, mc, "basic", json_JSET_basic_test)
 	runStep(t, mc, "geojson", json_JSET_geojson_test)
+	runStep(t, mc, "number", json_JSET_number_test)
+
 }
 func json_JSET_basic_test(mc *mockServer) error {
 	return mc.DoBatch([][]interface{}{
@@ -37,5 +39,18 @@ func json_JSET_geojson_test(mc *mockServer) error {
 		{"JSET", "mykey", "myid1", "properties.tags.-1", "hot"}, {"OK"},
 		{"JGET", "mykey", "myid1"}, {`{"type":"Feature","geometry":{"type":"Point","coordinates":[-115,44]},"properties":{"tags":["southwest","united states","hot"]}}`},
 		{"JDEL", "mykey", "myid1", "type"}, {"ERR missing type"},
+	})
+}
+
+func json_JSET_number_test(mc *mockServer) error {
+	return mc.DoBatch([][]interface{}{
+		{"JSET", "mykey", "myid1", "hello", "0"}, {"OK"},
+		{"JGET", "mykey", "myid1"}, {`{"hello":0}`},
+		{"JSET", "mykey", "myid1", "hello", "0123"}, {"OK"},
+		{"JGET", "mykey", "myid1"}, {`{"hello":"0123"}`},
+		{"JSET", "mykey", "myid1", "hello", "3.14"}, {"OK"},
+		{"JGET", "mykey", "myid1"}, {`{"hello":3.14}`},
+		{"JSET", "mykey", "myid1", "hello", "1.0e10"}, {"OK"},
+		{"JGET", "mykey", "myid1"}, {`{"hello":1.0e10}`},
 	})
 }

